src/Common/MemoryBufferWriter.cs

```diff
@@ -307,5 +313,27 @@ protected override void Dispose(bool disposing)
                 Reset();
             }
         }
+
+        /// <summary>
+        /// Holds a byte[] from the pool and a size value. Basically a Memory but guaranteed to be backed by an ArrayPool byte[], so that we know we can return it.
+        /// </summary>
+        private struct BufferSegment
```

Comment:
Why do we need this?

Reply:
Previously, the list was of "Full" segments. However, now a segment can be retired and put into the chain when only partially full (because there wasn't enough space in the buffer). We can't just track the array, we need to track how much of the array is used. I considered using `ReadOnlyMemory<byte>` in the list but that meant returning it to the pool required calling `TryGetArray`.

Reply:
> I considered using ReadOnlyMemory<byte> in the list but that meant returning it to the pool required calling TryGetArray.

Yea, I was wondering if this is better. Returning happens at the end and it might be more efficient than another struct.